tardis/plasma/properties/nlte_rate_equation_solver.py

```diff
@@ -0,0 +1,290 @@
+        atomic_numbers = (
+            rate_matrix_index.get_level_values(0).unique()[:-1].values
+        )  # dropping the n_e index
+        for atomic_number in atomic_numbers:
+            ion_numbers = rate_matrix.loc[atomic_number].index.get_level_values(
+                0
+            )
+            phi_block = phi_shell.loc[atomic_number]
+            rate_matrix_block = NLTERateEquationSolver.lte_rate_matrix_block(
+                phi_block, electron_density
+            )
+
+            nlte_ion_numbers = ion_numbers[
+                rate_matrix.loc[atomic_number].index.get_level_values(1)
+                == "nlte_ion"
+            ]
+            lte_ion_numbers = ion_numbers[
+                rate_matrix.loc[atomic_number].index.get_level_values(1)
+                == "lte_ion"
+            ]
+            for ion_number in nlte_ion_numbers:
+                rate_matrix_block = NLTERateEquationSolver.set_nlte_ion_rate(
+                    rate_matrix_block,
+                    atomic_number,
+                    ion_number,
+                    rad_recomb_rates.loc[(atomic_number,)],
+                    photo_ion_rate.loc[(atomic_number,)],
+                    coll_ion_rates.loc[(atomic_number,)],
+                    coll_recomb_rates.loc[(atomic_number,)],
+                )
+            rate_matrix.loc[
+                (atomic_number, slice(None)), (atomic_number)
+            ] = rate_matrix_block
+
+        last_row = NLTERateEquationSolver.prepare_last_row(atomic_numbers)
+        rate_matrix.loc[("n_e", slice(None))] = last_row
+        return rate_matrix
+
+    @staticmethod
+    def set_nlte_ion_rate(
+        rate_matrix_block,
+        atomic_number,
+        ion_number,
+        radiative_recombination_rate,
+        photo_ion_rates,
+        coll_ion_rate,
+        coll_recomb_rate,
+    ):
+        """Calculates the row for the species treated in NLTE ionization
+
+        Parameters
+        ----------
+        rate_matrix_block : numpy.array
+            The diagonal block corresponding to current atomic number.
+        atomic_number : int
+            Current atomic number
+        ion_number : int
+            Current ion number
+        radiative_recombination_rate : DataFrame
+            Rad. recomb. rate for current atomic number
+        photo_ion_rates : DataFrame
+            Photo ion. rate for current atomic number
+        coll_ion_rate : _type_
+            Coll. ion. rate for current atomic number
+        coll_recomb_rate : _type_
+            Coll. recomb. rate for current atomic number
+
+        Returns
+        -------
+        numpy.array
+            Rate matrix block with a changed row for NLTE ionization treatment
+        """
+        ion_rates = photo_ion_rates + coll_ion_rate
+        recomb_rate = radiative_recombination_rate + coll_recomb_rate
+        if atomic_number == ion_number:
+            rate_matrix_block[ion_number, :] = 1.0
```

Comment:
Is this not already done in the rate_matrix_block? It should also be needed for the LTE case. Maybe I am missing something?

Reply:
Done